src/main/scala/mesosphere/marathon/state/Group.scala

```diff
@@ -254,26 +254,59 @@ object Group {
   private def validPorts: Validator[Group] = {
     new Validator[Group] {
       override def apply(group: Group): Result = {
-        val groupViolations = group.transitiveApps.flatMap { app =>
-          val servicePorts = app.servicePorts.toSet
-
-          val ruleViolations = for {
-            existingApp <- group.transitiveApps
-            if existingApp.id != app.id // in case of an update, do not compare the app against itself
-            existingServicePort <- existingApp.servicePorts
-            if existingServicePort != 0 // ignore zero ports, which will be chosen at random
-            if servicePorts contains existingServicePort
-          } yield RuleViolation(
-            app.id,
-            s"Requested service port $existingServicePort conflicts with a service port in app ${existingApp.id}",
-            None)
-
-          if (ruleViolations.isEmpty) None
-          else Some(GroupViolation(app, "app contains conflicting ports", None, ruleViolations.toSet))
+        import collection.mutable
+        // Each service port should be unique across a cluster. We
+        // want to report conflicts if the same service port is used
+        // by multiple apps.
+        // We construct the mapping servicePort <- Set[AppKey]. This
+        // allows us to report any service port that has more than 1
+        // application using the same port.
+
+        // We keep track of the total number of ports to support an
+        // early exit condition.
+        var ports: Int = 0
+        var merged: mutable.MultiMap[Int, AppDefinition.AppKey] =
+          new mutable.HashMap[Int, mutable.Set[AppDefinition.AppKey]] with mutable.MultiMap[Int, AppDefinition.AppKey]
+
+        // Add each servicePort <- Application to the map.
+        for (app <- group.transitiveApps) {
+          for (port <- app.servicePorts) {
+            // We ignore randomly assigned ports identified by `0`.
+            if (port != 0) {
+              ports += 1
+              merged.addBinding(port, app.id)
+            }
+          }
         }
 
-        if (groupViolations.isEmpty) Success
-        else Failure(groupViolations.toSet)
+        // If the total number of unique ports is equal to the number
+        // of requested ports then we know there are no conflicts.
+        if (merged.size == ports) {
+          Success
+        } else {
+          // Otherwise we find all ports that have more than 1 app
+          // interested in them.
+          var groupViolations = Set.empty[RuleViolation]
+          for ((port, apps) <- merged) {
+            if (apps.size > 1) {
+              for (app <- apps; otherApp <- apps) {
+                // We report all conflicts (e.g. A -> B, and B -> A,
+                // but not A -> A or B -> B).
```

Comment:
If you had the HashMap be "HashMap[Int, mutable.ListBuffer]", we could report "A conflicts with itself". That, I'd consider a problem with the current edition.